File: core/mitta/api/http/memory.py
"""Memory CRUD, search and maintenance (API_DESIGN.md §3.2).

Every route reaches the engine through `MemoryService`, never through a
repository or the vector store directly. Bypassing the service is how a write
lands without waking the indexer, leaving a memory that exists but is not
searchable — a failure with no error and no log line.
"""

from __future__ import annotations

from fastapi import APIRouter, Query, Request, Response, status

from mitta.api.auth import RequireToken
from mitta.api.schemas.memory import (
    CreateMemoryRequest,
    MemoryListResponse,
    MemoryResource,
    MemoryStatsResponse,
    SearchHitResource,
    SearchRequest,
    SearchResponse,
    SweepResponse,
    UpdateMemoryRequest,
)
from mitta.memory.models import (
    MemoryDraft,
    MemoryKind,
    MemoryPatch,
    MemoryStatus,
    SourceKind,
)
from mitta.memory.service import MemoryService

router = APIRouter(prefix="/v1/memory", tags=["memory"])


def _service(request: Request) -> MemoryService:
    service: MemoryService = request.app.state.memory
    return service


@router.post(
    "",
    response_model=MemoryResource,
    status_code=status.HTTP_201_CREATED,
    summary="Store a memory",
)
async def create_memory(
    request: Request, body: CreateMemoryRequest, _: RequireToken
) -> MemoryResource:
    """Returns 201 even when the content already existed.

    `remember` is idempotent on content hash, so a duplicate returns the
    original rather than erroring. A 409 would be technically defensible and
    practically hostile: the caller asked for the fact to be remembered, and it
    is remembered.
    """
    draft = MemoryDraft(
        kind=body.kind,
        content=body.content,
        summary=body.summary,
        attributes=body.attributes,
        project_id=body.project_id,
        importance=body.importance,
        confidence=body.confidence,
        pinned=body.pinned,
        expires_at=body.expires_at,
        # Server-assigned. See CreateMemoryRequest for why the client cannot set it.
        source_kind=SourceKind.USER,
    )
    return MemoryResource.of(_service(request).remember(draft))


@router.get("", response_model=MemoryListResponse, summary="Browse memories")
async def list_memories(
    request: Request,
    _: RequireToken,
    kind: MemoryKind | None = None,
    project_id: str | None = None,
    memory_status: MemoryStatus = Query(default=MemoryStatus.ACTIVE, alias="status"),
    pinned_only: bool = False,
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
) -> MemoryListResponse:
    service = _service(request)
    memories = service.list_memories(
        kind=kind,
        project_id=project_id,
        status=memory_status,
        pinned_only=pinned_only,
        limit=limit,
        offset=offset,
    )
    return MemoryListResponse(
        memories=[MemoryResource.of(memory) for memory in memories],
        total=service.count(status=memory_status),
        limit=limit,
        offset=offset,
    )


@router.post("/search", response_model=SearchResponse, summary="Hybrid search")
async def search_memories(request: Request, body: SearchRequest, _: RequireToken) -> SearchResponse:
    """POST rather than GET.

    The query is conversational text — it belongs in a body, not in a URL that
    lands in history and access logs. R5 is about what leaves the machine, but
    the same instinct applies to what is written down locally.
    """
    service = _service(request)
    results = service.recall(
        body.query,
        limit=body.limit,
        project_id=body.project_id,
        semantic=body.semantic,
        keyword=body.keyword,
        touch=body.record_access,
    )
    index = service.index_status()
    return SearchResponse(
        query=body.query,
        hits=[SearchHitResource.of(result) for result in results],
        semantic_available=body.semantic and index.vector_count > 0,
    )


def _stats(request: Request) -> MemoryStatsResponse:
    service = _service(request)
    index = service.index_status()
    embedder = request.app.state.embedder
    descriptor = embedder.descriptor
    # `is_available` only exists on the downloadable provider. For the stand-in
    # the honest answer is that no weights were ever needed, so False here means
    # "the real model is not on disk" rather than "nothing works".
    downloaded = embedder.is_available() if hasattr(embedder, "is_available") else False
    return MemoryStatsResponse(
        active=service.count(),
        total=service.count(status=None),
        index_name=index.index_name,
        model_id=index.model_id,
        dim=index.dim,
        vectors_indexed=index.vector_count,
        pending_embeddings=service.pending_count(),
        index_consistent=index.consistent,
        embedding_degraded=descriptor.degraded,
        embedding_model_id=descriptor.id,
        embedding_model_downloaded=downloaded,
    )


@router.get("/stats", response_model=MemoryStatsResponse, summary="Engine health")
async def memory_stats(request: Request, _: RequireToken) -> MemoryStatsResponse:
    """Reports what is true, including when it is unflattering.

    `pending_embeddings` above zero and `embedding_degraded` true are both
    normal states the user is entitled to see, rather than a spinner that
    implies work is happening when it cannot.
    """
    return _stats(request)


@router.get("/{memory_id}", response_model=MemoryResource, summary="Read one memory")
async def get_memory(request: Request, memory_id: str, _: RequireToken) -> MemoryResource:
    return MemoryResource.of(_service(request).get(memory_id))


@router.patch("/{memory_id}", response_model=MemoryResource, summary="Edit a memory")
async def update_memory(
    request: Request, memory_id: str, body: UpdateMemoryRequest, _: RequireToken
) -> MemoryResource:
    patch = MemoryPatch.model_validate(body.model_dump(exclude_unset=True))
    return MemoryResource.of(_service(request).update(memory_id, patch))


@router.post(
    "/{memory_id}/correct", response_model=MemoryResource, summary="Supersede with a correction"
)
async def correct_memory(
    request: Request, memory_id: str, body: CreateMemoryRequest, _: RequireToken
) -> MemoryResource:
    """The original is kept and linked, never overwritten.

    "I moved to Bangalore" does not make "I lived in Hyderabad" false — it makes
    it historical, and only one of those is answerable if the row is gone.
    """
    draft = MemoryDraft(
        kind=body.kind,
        content=body.content,
        summary=body.summary,
        attributes=body.attributes,
        project_id=body.project_id,
        importance=body.importance,
        confidence=body.confidence,
        pinned=body.pinned,
        expires_at=body.expires_at,
        source_kind=SourceKind.USER,
    )
    return MemoryResource.of(_service(request).correct(memory_id, draft))


@router.post("/{memory_id}/forget", response_model=MemoryResource, summary="Demote to forgotten")
async def forget_memory(request: Request, memory_id: str, _: RequireToken) -> MemoryResource:
    """Reversible. The row survives; only its visibility changes."""
    return MemoryResource.of(_service(request).forget(memory_id))


@router.post("/{memory_id}/restore", response_model=MemoryResource, summary="Undo a forget")
async def restore_memory(request: Request, memory_id: str, _: RequireToken) -> MemoryResource:
    return MemoryResource.of(_service(request).restore(memory_id))


@router.delete("/{memory_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Delete permanently")
async def purge_memory(request: Request, memory_id: str, _: RequireToken) -> Response:
    """Irreversible, and the only endpoint that destroys data.

    Deliberately a separate verb from `forget`: a UI that maps "delete" onto
    this without a confirmation is doing something the user cannot undo, and the
    HTTP method should make that obvious to whoever wires it up.
    """
    _service(request).purge(memory_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/maintenance/sweep", response_model=SweepResponse, summary="Run a decay sweep")
async def sweep(request: Request, _: RequireToken) -> SweepResponse:
    report = _service(request).sweep()
    return SweepResponse(examined=report.examined, expired=report.expired, decayed=report.decayed)


@router.post("/maintenance/reindex", response_model=MemoryStatsResponse, summary="Rebuild vectors")
async def reindex(request: Request, _: RequireToken) -> MemoryStatsResponse:
    """Drops every vector and re-embeds from SQLite.

    A failure here propagates rather than being caught: a reindex that cannot
    run because the model is absent must say so, not report success over an
    index it silently left empty.
    """
    _service(request).reindex()
    return _stats(request)

File: apps/desktop/src/state/memory.ts
/**
 * Memory surface state.
 *
 * Kept separate from the main store slice on purpose: this is *server* state —
 * a cache of rows the sidecar owns — and conflating it with UI state is how a
 * stale copy ends up being treated as the truth. Every mutation here re-reads
 * from the server rather than patching the local list optimistically, because a
 * memory the user believes they pinned but did not is worse than a spinner.
 */

import { create } from 'zustand';

import {
  type ApiClient,
  ApiError,
  type Memory,
  type MemoryKind,
  type MemoryStats,
  type MemoryStatus,
  type SearchHit,
} from '@/lib/api/client';

export type MemoryView = 'browse' | 'search';

export interface MemoryFilters {
  kind: MemoryKind | 'all';
  status: MemoryStatus;
  pinnedOnly: boolean;
}

interface MemoryState {
  client: ApiClient | null;
  view: MemoryView;
  filters: MemoryFilters;

  memories: Memory[];
  total: number;
  hits: SearchHit[];
  /** Null when the last search was keyword-only or nothing was indexed. */
  semanticAvailable: boolean | null;
  stats: MemoryStats | null;

  query: string;
  selectedId: string | null;
  loading: boolean;
  error: string | null;

  attach: (client: ApiClient | null) => void;
  setView: (view: MemoryView) => void;
  setFilters: (patch: Partial<MemoryFilters>) => void;
  setQuery: (query: string) => void;
  select: (id: string | null) => void;

  refresh: () => Promise<void>;
  search: (query: string, options?: { recordAccess?: boolean }) => Promise<void>;
  create: (content: string, kind: MemoryKind, pinned: boolean) => Promise<void>;
  setPinned: (id: string, pinned: boolean) => Promise<void>;
  forget: (id: string) => Promise<void>;
  restore: (id: string) => Promise<void>;
  purge: (id: string) => Promise<void>;
  reindex: () => Promise<void>;
}

const DEFAULT_FILTERS: MemoryFilters = { kind: 'all', status: 'active', pinnedOnly: false };

function describe(error: unknown): string {
  if (error instanceof ApiError) return error.message;
  if (error instanceof Error) return error.message;
  return String(error);
}

export const useMemoryStore = create<MemoryState>((set, get) => ({
  client: null,
  view: 'browse',
  filters: DEFAULT_FILTERS,

  memories: [],
  total: 0,
  hits: [],
  semanticAvailable: null,
  stats: null,

  query: '',
  selectedId: null,
  loading: false,
  error: null,

  attach: (client) => {
    set({ client });
    if (client !== null) void get().refresh();
  },

  setView: (view) => set({ view }),

  setFilters: (patch) => {
    set({ filters: { ...get().filters, ...patch } });
    void get().refresh();
  },

  setQuery: (query) => set({ query }),
  select: (selectedId) => set({ selectedId }),

  refresh: async () => {
    const { client, filters } = get();
    if (client === null) return;

    set({ loading: true, error: null });
    try {
      const [list, stats] = await Promise.all([
        client.listMemories({
          ...(filters.kind === 'all' ? {} : { kind: filters.kind }),
          status: filters.status,
          pinnedOnly: filters.pinnedOnly,
          limit: 100,
        }),
        client.memoryStats(),
      ]);
      set({ memories: list.memories, total: list.total, stats, loading: false });
    } catch (error) {
      set({ error: describe(error), loading: false });
    }
  },

  search: async (query, options = {}) => {
    const { client } = get();
    if (client === null) return;

    const trimmed = query.trim();
    if (trimmed.length === 0) {
      set({ hits: [], semanticAvailable: null, view: 'browse' });
      return;
    }

    set({ loading: true, error: null, view: 'search' });
    try {
      const response = await client.searchMemories({
        query: trimmed,
        limit: 25,
        semantic: true,
        keyword: true,
        // Typing in the search box is not using a memory. Recording access here
        // would let idle browsing keep trivia alive at the expense of memories
        // that are genuinely consulted.
        record_access: options.recordAccess ?? false,
      });
      set({
        hits: response.hits,
        semanticAvailable: response.semantic_available,
        loading: false,
      });
    } catch (error) {
      set({ error: describe(error), loading: false, hits: [] });
    }
  },

  create: async (content, kind, pinned) => {
    const { client } = get();
    if (client === null) return;
    set({ error: null });
    try {
      await client.createMemory({ content, kind, pinned, importance: 0.5, confidence: 1.0 });
      await get().refresh();
    } catch (error) {
      set({ error: describe(error) });
    }
  },

  setPinned: async (id, pinned) => {
    await mutate(get, set, (client) => client.updateMemory(id, { pinned }));
  },

  forget: async (id) => {
    await mutate(get, set, (client) => client.forgetMemory(id));
  },

  restore: async (id) => {
    await mutate(get, set, (client) => client.restoreMemory(id));
  },

  purge: async (id) => {
    await mutate(get, set, (client) => client.purgeMemory(id));
    if (get().selectedId === id) set({ selectedId: null });
  },

  reindex: async () => {
    const { client } = get();
    if (client === null) return;
    set({ loading: true, error: null });
    try {
      const stats = await client.reindexMemories();
      set({ stats, loading: false });
      await get().refresh();
    } catch (error) {
      set({ error: describe(error), loading: false });
    }
  },
}));

/**
 * Run a mutation, then re-read. Never patches the local row optimistically:
 * the server applies rules the client does not know — a pinned memory refuses
 * to be forgotten, and a `forget` that silently no-ops would leave the UI
 * showing a state the database never entered.
 */
async function mutate(
  get: () => MemoryState,
  set: (patch: Partial<MemoryState>) => void,
  action: (client: ApiClient) => Promise<unknown>,
): Promise<void> {
  const { client } = get();
  if (client === null) return;
  set({ error: null });
  try {
    await action(client);
    await get().refresh();
    if (get().view === 'search') await get().search(get().query);
  } catch (error) {
    set({ error: describe(error) });
  }
}
